File: 2023/Arrays/bestTimeToBuyAndSellStock.py
# You are given an array prices where prices[i] is the price of a given stock on the ith day.

# You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

# Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

# Example 1:
# Input: prices = [7,1,5,3,6,4]
# Output: 5
# Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
# Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.

def maxProfit(nums):
    min_value = max_value = nums[0]
    max_profit = 0 
    for i in nums[1:]:
        if i < min_value:
            min_value = i
            max_value = i
        else:
            max_value = max(i, max_value)
            max_profit = max(max_profit, max_value - min_value)
    return max_profit

print(maxProfit([7,1,5,3,6,4]))
print(maxProfit([7,6,4,3,1]))

File: 2023/Arrays/bestTimeToBuyAndSellStock.java
class BestTimeBuySellStock {
    public static int maxProfit(int[] nums) {
        int minVal, maxVal, maxProfit;
        minVal = maxVal = nums[0];
        maxProfit = 0;
        for (int i=1; i<nums.length; i++) {
            if (nums[i] < minVal) {
                minVal = nums[i];
                maxVal = nums[i];
            }
            else {
                maxVal = Math.max(maxVal, nums[i]);
            }
            maxProfit = Math.max(maxProfit, maxVal - minVal);
        }
        return maxProfit;
    }

    public static void main(String[] args) {
        int[] prices = {7,1,5,3,6,4};
        System.out.println(maxProfit(prices));
    }
}